### ISSUE TITLE
Allow naming files as dependencies.

### ISSUE THREAD

**Issue by brownhead** [brownhead/phial#65]
Right now it's possible to do this with pages and pipelines by just writing code like:

```python
@phial.page(["a.txt", "b.txt"])
def foo(...):
    ...
```

But this way of declaring files as dependencies also makes other things happen (your page function will be called two times in this example for instance). To ground things in practicalness, here's a page function from my blog:

```python
@phial.page(["posts/*.rst", "drafts/*.rst"])
def post_page(source_file):
    if not os.environ.get("DRAFTING") and source_file.name.startswith("drafts/"):
        return None

    template = phial.open_file("posts/template.htm").read()

    frontmatter, content = phial.parse_frontmatter(source_file)

    # Use docutils to render the restructured text
    post_body = render_rst(content.read())

    # Use mustache to plug everything into the template
    renderer = pystache.Renderer()
    content = renderer.render(template, frontmatter, {"body": post_body})

    return phial.file(
        name=phial.swap_extension(source_file.name, ".htm"),
        content=content,
        metadata=frontmatter or {})
```

The important bit to note is the line `template = phial.open_file("posts/template.htm").read()`. Here I'm reading from a template file. This dependency isn't declared, and if Phial was smarter about reloading things (and I intend for it to be soon), it wouldn't know to call into this page function again.

A naive solution to this raises some problems though. If I simply allowed users to declare files as dependencies with no further flexibility, things could get unpleasant if the dependencies changed per file (which seems reasonable). It might not be necessary to consider this edge-y case thoroughly right now though as long as I can convince myself that the solution I come up with won't prohibit handling this case later on.

**Comment by brownhead** [comment]
The `depends_on` parameter already exists. But it seems a little weird to just include files in there as well:

```python
@phial.page(["posts/*.rst"], depends_on=["posts/template.htm"])
def bla()...
```

This makes me think that if `posts/template.htm` isn't available, this page function won't execute. That might be what I want though... This could cause a warning to be logged, but would otherwise be fine.

**Comment by brownhead** [comment]
Hmm, this also makes me think of make files. This is problematic because Phial has a strong concept of some kind of source tree that is separate from its concept of the build/output tree. If I could put files in `depends_on`, I would expect to be able to put a file created by another task and have it figure things out from there (like make does).

**Comment by brownhead** [comment]
Let's say I am a task. I may depend on other tasks, which mean that those tasks must be complete in order for me to run. There is a separate but related concept of sources that affect how I run and what I create.

So if I were to isolate each responsibility and make a number of attributes of a task, each a list, what would I get?

* `dependent_tasks`: Tasks that must be complete before I run.
* `source_tasks`: Tasks that I use the output of or side-effects of (we'll probably have to restrict ourselves to output since we won't be able to detect all side-effects), so these must also be `dependent_tasks`, and I will be out of date if one of my source tasks is run and produces different output.
* `source_files`: Files that I use to create my output. I will be out of date if my list of source files changes (like if it's a glob and a file is deleted or added) or if any of my source files changes (metadata or content).

If a listed source doesn't exist, I probably can't run... So it seems like `dependent_tasks` and `source_tasks` shouldn't ever be different. Why would I ever just require a task to be completed while not actually caring what it creates... It seems weird.

**Comment by brownhead** [comment]
So it seems like I only have sources because while the "dependency" distinction is separate (it could be different and things might behave how you might expect), practically the lists are never different.

What are the characteristics of a source? It can be invalid (so doesn't exist, or raised an error while running), it can change (requiring dependents to be rerun), it can be complete or incomplete...

From the "thing that runs the tasks"'s point of view, there's probably just the states "invalid", "incomplete", and "complete". Because change would just mark the task as incomplete.

**Comment by brownhead** [comment]
The page decorator can use "additional_sources" to let users mark things like template files. It might also make sense to do away with the page functions since it's really only saving the user from having to write a for loop (if it was just a function that got executed rather than a function that got executed once for each source).

I also had an idea yesterday that I forgot to write down. The sources could probably (optionally) be a dict like so:

```
{
    "css_files": [...],
    "templates": [...],
}
```

Then you could inspect the dict from within the task. This would just save people from having to put the template file names in a variable... Maybe not super useful though. Probably shouldn't be done outright, but might be a solution to a problem I've yet to encounter.
